src/protocol/attribute.rs

```diff
@@ -0,0 +1,224 @@
+    #[allow(dead_code)]
+    pub async fn execute<M, F>(
+        &self,
+        input: &'a [Vec<Replicated<F>>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+        F: Field,
+    {
+        match self.step {
+            AttributionModel::LastTouch => self.last_touch_attribution(input).await,
+        }
+    }
+
+    async fn last_touch_attribution<M, F>(
+        &self,
+        input: &'a [Vec<Replicated<F>>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+        F: Field,
+    {
+        // TOOD: input length range check. Can we assume `assert!(num_rows < u32::MAX)`?
+        let num_rows = input[0].len();
+
+        // 1. Create Credit and Stop bit vectors
+        let mut credits: Vec<Replicated<F>> = Vec::with_capacity(num_rows);
+        input[Self::_VALUE].iter().for_each(|x| credits.push(*x));
+
+        let one = self.ctx.gateway.get_channel(self.step).share_of_one::<F>();
+        let mut stop_bits: Vec<Replicated<F>> = vec![one; num_rows];
+
+        // 2. Accumulate
+        //
+        // Last touch accumulation algorithm:
+        //     b = stop_bit * successor.helper_bit * successor.trigger_bit;
+        //     credit = credit + b * successor.credit;
+        //     stop_bit = b * successor.stop_bit;
+        let mut step_size = 1;
+        while step_size < num_rows {
+            let end = num_rows - step_size;
+
+            // first, calculate [successor.helper_bit * successor.trigger_bit]
+            let mut b = self
+                .multiply_all(
+                    &input[Self::_HELPER_BIT][step_size..],
+                    &input[Self::_IS_TRIGGER_BIT][step_size..],
+                )
+                .await?;
+
+            // since `stop_bits` is initialized with `1`s, we only multiply [stop_bit] on the second and later iterations
+            if step_size > 1 {
+                b = self.multiply_all(&stop_bits[..end], &b).await?;
+            }
+
+            // [b * successor.credit]
+            let b_cred = self.multiply_all(&b[..end], &credits[step_size..]).await?;
+
+            // [credit] = [credit + b * successor.credit]
+            b_cred
+                .into_iter()
+                .enumerate()
+                .for_each(|(i, x)| credits[i] += x);
+
+            // [stop_bit] = [b * successor.stop_bit]
+            // for the same reason as calculating [b], we skip the multiplication on the first iteration
+            let b_stop = if step_size > 1 {
+                self.multiply_all(&b[..end], &stop_bits[step_size..])
+                    .await?
+            } else {
+                b.clone()
+            };
+
+            b_stop
+                .into_iter()
+                .enumerate()
+                .for_each(|(i, x)| stop_bits[i] = x);
+
+            step_size *= 2;
+        }
+
+        Ok(credits)
+    }
+
+    async fn multiply_all<M, F>(
```

Comment:
Is this a function that we should be providing on the Gateway?

Reply:
Makes sense. Many protocols seem to need this, so let's put it in the gateway.

Reply:
@richajaindce could have used this in GenBitPerm